src/internal/queues.mjs

```diff
@@ -44,19 +29,34 @@ export class Queue {
   }
 }
 
-export class Exchange extends Queue {
+class Consumer {
+  constructor (queueItem) {
+    this.queueItem = queueItem
+  }
+  isEmpty () {
+    return !this.queueItem.previous
+  }
   shift () {
-    throw new Error('Unsupported')
+    if (this.isEmpty()) throw new Error('Queue is empty')
+    const data = this.queueItem.previous.data
+    this.queueItem = this.queueItem.previous
+    return data
+  }
+  clone () {
+    return new Consumer(this.queueItem)
   }
+}
 
-  spawnConsumer () {
-    if (!this.head) throw new Error('You cannot spawn a new consumer after setting calling noMoreConsumers')
-    return new Consumer(this.head)
+export class Exchange extends Queue {
+  shift () {
+    throw new Error('Unsupported')
   }
 
-  noMoreConsumers () {
-    // this enables to garbage collect all the consumed items
+  getConsumer () {
```

Comment:
I think with the change to consumers cloning each other, the exchange base class no longer makes much sense. What about just having it be a function used like this: `const { consumer, cacheValue /* fn(value) */ } = createExchange();`?

Reply:
Actually no that won't work because you still need `noMoreConsumers()` to be part of the API, which you have gotten rid of here. We put that in for forking infinite iterables. You can't unallocate the exchange because you still need to be able to push items into it. Exchange also can't keep its head, because that forces the GC to keep everything in the linked list, even items that have been consumed by every consumer.

Reply:
Ignore my previous comment, it was wholly wrong.

Reply:
Ah damn, no, it wasn't wholly wrong. This breaks fork. We need to add a new test to fork. The case which breaks is:
```js
forks = fork(range())
forks.next().value.next() // 1
forks.next().value.next() // 2, but should be 1.
```

Reply:
Wait, I ran that as a test on your branch and it passed. My head hurts.
It passes because you are keeping the original consumer to clone from, which mean that the bug is far more subtle. Keeping that consumer means that you are preventing any cache data from ever being GC'd.

Reply:
This can't be done because, as I explain [below](https://github.com/sithmel/iter-tools/pull/208#issuecomment-471233351) I need to create the consumer in a specific place, and not necessarily together in the same scope as the main object.